### ISSUE TITLE
Providing default docker image with custom built AMI

### ISSUE THREAD

**Issue by manniche** [bottlerocket-os/bottlerocket#1128]
**Platform I'm building on:**
Linux 5.8.9-arch2-1 GNU/Linux

**What I expected to happen:**
Find information on how to provide a default image to package with Bottlerocket OS when doing a custom build

I would like to be able to include a default image with the Bottlerocket OS so that upon starting the OS, and subsequently trying to start a docker container, I would not have to wait for the image being pulled, as is customary in e.g. Fargate tasks (see https://github.com/aws/containers-roadmap/issues/358)

**What actually happened:**
Found information on how to build a custom Bottlerocket OS in https://github.com/bottlerocket-os/bottlerocket/blob/develop/BUILDING.md but not information on how to provide a default docker image

**How to reproduce the problem:**
N/A

**Comment by samuelkarp** [comment]
Hi @manniche, thanks for opening this issue!  I don't believe we have an existing mechanism to facilitate this use-case today, but I can see how it would be useful to have support for cached container images.  Can you tell me a bit more about what you're looking for?  Are you using one of our Kubernetes variants, our ECS variant, or the `aws-dev` variant?

**Comment by samuelkarp** [comment]
Here's an idea for a work-around that you could use today:  On EC2, the Bottlerocket AMIs are configured with two EBS volumes: one attached at `/dev/xvda` that contains the system software and a second attached at `/dev/xvdb` available for persistent storage (including container images).  If you launch an instance, pull images to that instance, and create a snapshot of the volume attached at `/dev/xvdb`, you can substitute the snapshot you created in the `BlockDeviceMappings` field (overriding the default provided for `/dev/xvdb`) when you launch new instances afterward.

**Comment by manniche** [comment]
Thanks for the suggestion @samuelkarp. I'll try to test it out but would also like to have a fully automated process with not too many hacks to build out an AMI with cached images on them.

**Comment by samuelkarp** [comment]
> I'll try to test it out but would also like to have a fully automated process with not too many hacks to build out an AMI with cached images on them.

Absolutely!  I'll re-open this issue and leave it open to track the request for a more complete design.  (If we do end up deciding that the approach I outlined is the one we want people to use, I'd like to have us document it somewhere other than an issue comment too :smile:.)

**Comment by pabloxio** [comment]
When I was using the AL2 AMI I added something like this at the end of the instance user_data:
```bash
....
$(aws ecr get-login --no-include-email --region us-west-2)
docker pull MY_LARGE_ECR_DOCKER_IMAGE
```

@samuelkarp What about having an additional option like this:

```toml
[settings.ecr.cache]
images = [ "MY_LARGE_ECR_DOCKER_IMAGE" ]
```
And bottlerocket just pull those images at instance creation time?

**Comment by zmrow** [comment]
Thanks for the idea @pabloxio and I like where your head is at!  It's a cool idea and definitely something to consider.  

Unfortunately, it doesn't solve the issue of bundling a container image with the OS.  A user would still need to wait for the initial pull at runtime.  I do think there's something to this "cache" idea, however we'd probably want it to be a bit more generic over container registry.

@samuelkarp 's suggested solution above is still the best option here.

**Comment by pat-s** [comment]
@samuelkarp Thanks for the great tip, I used it just now and it works great! 

I would also vote for the (documented) option to build a custom AMI with additional images backed in - this would simplify updating tasks. This is already possible in [amazon-eks-ami](https://github.com/awslabs/amazon-eks-ami), see also [this outdated PR](https://github.com/awslabs/amazon-eks-ami/pull/677) for inspiration.

Curretnly one needs to go "the extra mile" and for every new image 

1. pull it once
2. create a snapshot
3. update the `block_device_mappings` setting with the new `snapshot_id`

For my use case this option is really important as it is a ~3GB (compressed) image which would otherwise take > 3 min to pull+unpack. Having it available within the node image is essential.

**Comment by andrescaroc** [comment]
This feature is needed. 

In the meantime, I was trying to follow the tip from @samuelkarp, but I am a little bit lost. Can you give me more hints about the process? 

- I understand that I should start an EC2 instance with bottlerocket AMI (I choose `bottlerocket-aws-k8s-1.23-x86_64-v1.11.1-104f8e0f`).
- This EC2 instance has the IAM profile with the `AmazonSSMManagedInstanceCore`  and `AmazonEC2ContainerRegistryReadOnly` permisions on its IAM role.
- I can connect to the EC2 instance through aws SSM Session Manager, then I run `enter-admin-container`, then I run `sheltie` and now I can pull an image with the `ctr` cli tool that is included. 

Is there a way to do it in a more automated way? something that can be used in a CI pipeline every time my team release a new container image?

EDIT: I noticed that the pulled image is stored in `/var/lib/...` but from the documentation I understand that it should be in `/local/var/..`  to be persistent in the second ebs `/dev/xvdb`. I suppose this due to the fact that the admin container runs over a different containerd that kubernetes.
```
ls /var/lib/containerd/io.containerd.content.v1.content/blobs/sha256/
060bfa6be...  
41d8806bd...  
b34d5ba6fa...
0a208d30b...  
44d36245a8... 
cf81dddbbd...
106a8e54d5...
```
@pat-s How do you make the pulled image available in the second ebs before creating the snapshot?

EDIT 2: From the  admin container you can pull the image and make it available for the containerd that will use the kubelet using the following command:
```shell
ctr -a /run/containerd/containerd.sock -n k8s.io images pull -u <user>:<password> <image>:<tag>
```
Beware that the containerd socket address could be different in you case and depends on the variant you use. My example works for `aws-k8s-1.2X` >= 1.23

It works pretty well
I hope this helps someone else trying this approach.

**Comment by omegayal** [comment]
@andrescaroc 
> 
> ```shell
> ctr -a /run/containerd/containerd.sock -n k8s.io images pull -u <user>:<password> <image>:<tag>
> ```
> 
> Beware that the containerd socket address could be different in you case and depends on the variant you use. My example works for `aws-k8s-1.2X` >= 1.23
> 
> It works pretty well I hope this helps someone else trying this approach.

I am trying to follow your instructions to pull a custom image with the bottlerocket-aws-ecs ami. However, the above command seems to work only for k8s ami. Do you happen to know the command for ecs ami? Thanks.

**Comment by andrescaroc** [comment]
> 
> I am trying to follow your instructions to pull a custom image with the bottlerocket-aws-ecs ami. However, the above command seems to work only for k8s ami. Do you happen to know the command for ecs ami? Thanks.

To be honest  I don't know, but maybe you have to change the namespace of the containerd, i mean the `-n k8s.io` section of the command.

I have never worked with ECS so I don't know how to guide you. I am sorry.

**Comment by d3vr4ndOm** [comment]
Hello,

I am also currently exploring options on how speed up EKS managed node group workers and thought it would be as simple as having a custom AMI eg. using packer and pull the images... But it looks like it is not that trivial.

From what I understood is that the custom data is stored on a second EBS volume mounted to the instance. But this volume also has to be pulled? At least this seems to be the case for AL2 instance: https://github.com/awslabs/amazon-eks-ami/issues/1273#issuecomment-1540799080

For the Bottlerocket image they linked towards this mechanism: https://github.com/aws-samples/containers-blog-maelstrom/tree/main/bottlerocket-images-cache

My question here would be why it should be faster for the bottlerocket instance compared to the AL2 instance. This would mean that the mechanism for the bottlerocket image/EBS volume is somehow different? Can anybody explain what the differenceis  and if the boot time can really be speed up?

Thanks!
